### ISSUE TITLE
Can't get proxy to work with gRPC server

### ISSUE THREAD

**Issue by bazaglia** [ahmetb/runsd#48]
I have both a gRPC server and client running on Cloud Run and the communication between the two of them works fine when calling `my-service-name-hash-ez.run.app:443` and providing to the client `grpc.credentials.createSsl()` – that's because Cloud Run exposes the public service under HTTPS.

When I make them private and trying to add runsd to handle the authentication, calling `my-service-name:80` this time, I get the error:

```
2022-10-06 20:01:52.298 CESTD 2022-10-06T18:01:51.898Z | subchannel | (21) ::1:80 connection closed with error 68968495794112:error:1408F10B:SSL routines:ssl3_get_record:wrong version number:../deps/openssl/openssl/ssl/record/ssl3_record.c:332:
```

If I use `grpc.credentials.createInsecure()` instead, since now the request is made to :80 instead of :443 so it gets proxied, I get the error:

```
http: proxy error: x509: certificate signed by unknown authority
```

I've already verified that the receiving Cloud Run service has the principal of the caller for the role Cloud Run Invoker.

**Comment by ahmetb** [comment]
I think you need to make grpc use h2c. It defaults to http. I think insecure option only skips server TLS validation but it still uses TLS (the error says so). 

Maybe looking at logs around the error timeframe can reveal some stuff.

**Comment by bazaglia** [comment]
I tried enabling h2c on the Cloud Run service that runs the gRPC server since it wasn't enabled by default - https://cloud.google.com/run/docs/configuring/http2

I have the gRPC Node.js client making an insecure request to `my-service-name:80`. It does http2 over cleartext when I provide the insecure credentials according to this issue: https://github.com/grpc/grpc-node/issues/1492#issuecomment-654319967

I verify it's proxying, and the URL is being properly rewritten, but still get the error `http: proxy error: x509: certificate signed by unknown authority`, and that's happening in the proxy call as I understand from the logs. Do you have any thoughts on that?

I have both runsd verbose logging and gRPC client debug mode enabled therefore got as much log info as I could get:

```
2022-10-07 09:43:33.890 CESTD 2022-10-07T07:43:33.890Z | channel | (1) dns:my-service-name:80 createCall [1] method="/MyService/SomeGrpcMethod", deadline=Infinity
2022-10-07 09:43:33.890 CESTD 2022-10-07T07:43:33.890Z | call_stream | [1] Sending metadata
2022-10-07 09:43:33.890 CESTD 2022-10-07T07:43:33.890Z | dns_resolver | Looking up DNS hostname my-service-name
2022-10-07 09:43:33.890 CESTD 2022-10-07T07:43:33.890Z | resolving_load_balancer | dns:my-service-name:80 IDLE -> CONNECTING
2022-10-07 09:43:33.890 CESTD 2022-10-07T07:43:33.890Z | connectivity_state | (1) dns:my-service-name:80 IDLE -> CONNECTING
2022-10-07 09:43:33.891 CESTD 2022-10-07T07:43:33.890Z | channel | (1) dns:my-service-name:80 callRefTimer.ref | configSelectionQueue.length=1 pickQueue.length=0
2022-10-07 09:43:33.891 CESTI1007 07:43:33.891128 1 dns.go:51] [dns] > Q0: type=AAAA name=my-service-name.europe-west4.run.internal.
2022-10-07 09:43:33.891 CESTI1007 07:43:33.891159 1 dns.go:123] [dns] < MATCH type=AAAA name=my-service-name.europe-west4.run.internal.
2022-10-07 09:43:33.891 CESTI1007 07:43:33.891223 1 dns.go:51] [dns] > Q0: type=TXT name=my-service-name.
2022-10-07 09:43:33.891 CESTI1007 07:43:33.891234 1 dns.go:154] [dns] >> recursing type=TXT name=my-service-name.
2022-10-07 09:43:33.891 CESTI1007 07:43:33.891272 1 dns.go:51] [dns] > Q0: type=A name=my-service-name.europe-west4.run.internal.
2022-10-07 09:43:33.891 CESTI1007 07:43:33.891300 1 dns.go:123] [dns] < MATCH type=A name=my-service-name.europe-west4.run.internal.
2022-10-07 09:43:33.892 CESTD 2022-10-07T07:43:33.892Z | call_stream | [1] write() called with message of length 32
2022-10-07 09:43:33.892 CESTD 2022-10-07T07:43:33.892Z | call_stream | [1] end() called
2022-10-07 09:43:33.893 CESTD 2022-10-07T07:43:33.893Z | call_stream | [1] deferring writing data chunk of length 37
2022-10-07 09:43:33.900 CESTD 2022-10-07T07:43:33.900Z | dns_resolver | Resolved addresses for target dns:my-service-name:80: [::1:80,127.0.0.1:80]
2022-10-07 09:43:33.900 CESTD 2022-10-07T07:43:33.900Z | pick_first | Connect to address list ::1:80,127.0.0.1:80
2022-10-07 09:43:33.900 CESTD 2022-10-07T07:43:33.900Z | subchannel | (9) ::1:80 Subchannel constructed with options {}
2022-10-07 09:43:33.900 CESTD 2022-10-07T07:43:33.900Z | subchannel_refcount | (9) ::1:80 refcount 0 -> 1
2022-10-07 09:43:33.903 CESTD 2022-10-07T07:43:33.903Z | subchannel_refcount | (10) 127.0.0.1:80 refcount 3 -> 2
2022-10-07 09:43:33.904 CESTD 2022-10-07T07:43:33.904Z | call_stream | Starting stream [1] on subchannel (10) 127.0.0.1:80 with headers
2022-10-07 09:43:33.904 CEST traceparent: 00-f68829e5f1b82b74cbd0fd99e75fcfba-b7b284647f548aee-01
2022-10-07 09:43:33.904 CEST grpc-accept-encoding: identity,deflate,gzip
2022-10-07 09:43:33.904 CEST accept-encoding: identity
2022-10-07 09:43:33.904 CEST :authority: my-service-name:80
2022-10-07 09:43:33.904 CEST user-agent: grpc-node-js/1.6.7
2022-10-07 09:43:33.904 CEST content-type: application/grpc
2022-10-07 09:43:33.904 CEST :method: POST
2022-10-07 09:43:33.904 CEST :path: /MyService/SomeGrpcMethod
2022-10-07 09:43:33.904 CEST te: trailers
2022-10-07 09:43:33.904 CESTD 2022-10-07T07:43:33.904Z | subchannel_flowctrl | (10) 127.0.0.1:80 local window size: 65535 remote window size: 65535
2022-10-07 09:43:33.904 CESTD 2022-10-07T07:43:33.904Z | subchannel_internals | (10) 127.0.0.1:80 session.closed=false session.destroyed=false session.socket.destroyed=false
2022-10-07 09:43:33.904 CESTD 2022-10-07T07:43:33.904Z | call_stream | [1] attachHttp2Stream from subchannel 127.0.0.1:80
2022-10-07 09:43:33.904 CESTD 2022-10-07T07:43:33.904Z | subchannel_refcount | (10) 127.0.0.1:80 callRefcount 0 -> 1
2022-10-07 09:43:33.904 CESTD 2022-10-07T07:43:33.904Z | call_stream | [1] sending data chunk of length 37 (deferred)
2022-10-07 09:43:33.904 CESTD 2022-10-07T07:43:33.904Z | call_stream | [1] calling end() on HTTP/2 stream
2022-10-07 09:43:33.904 CESTD 2022-10-07T07:43:33.904Z | subchannel | (9) ::1:80 CONNECTING -> READY
2022-10-07 09:43:33.905 CESTI1007 07:43:33.905173 1 proxy.go:59] [director] receive req host=my-service-name:80
2022-10-07 09:43:33.905 CESTI1007 07:43:33.905195 1 proxy.go:84] [director] rewrote host=my-service-name to=my-service-name-qwsbralb2c-ez.a.run.app new_url="https://my-service-name-qwsbralb2c-ez.a.run.app/MyService/SomeGrpcMethod"
2022-10-07 09:43:33.905 CESTI1007 07:43:33.905247 1 proxytransport.go:77] [proxy] start: POST url=https://my-service-name-qwsbralb2c-ez.a.run.app/MyService/SomeGrpcMethod
2022-10-07 09:43:33.905 CESTD 2022-10-07T07:43:33.905Z | subchannel | (10) 127.0.0.1:80 new settings received: {"headerTableSize":4096,"enablePush":true,"initialWindowSize":1048576,"maxFrameSize":1048576,"maxConcurrentStreams":250,"maxHeaderListSize":1048896,"maxHeaderSize":1048896,"enableConnectProtocol":false}
2022-10-07 09:43:33.905 CESTD 2022-10-07T07:43:33.905Z | subchannel | (10) 127.0.0.1:80 local settings acknowledged by remote: {"headerTableSize":4096,"enablePush":true,"initialWindowSize":65535,"maxFrameSize":16384,"maxConcurrentStreams":4294967295,"maxHeaderListSize":4294967295,"maxHeaderSize":4294967295,"enableConnectProtocol":false}
2022-10-07 09:43:33.905 CESTD 2022-10-07T07:43:33.905Z | subchannel | (9) ::1:80 new settings received: {"headerTableSize":4096,"enablePush":true,"initialWindowSize":1048576,"maxFrameSize":1048576,"maxConcurrentStreams":250,"maxHeaderListSize":1048896,"maxHeaderSize":1048896,"enableConnectProtocol":false}
2022-10-07 09:43:33.905 CESTD 2022-10-07T07:43:33.905Z | subchannel | (9) ::1:80 local settings acknowledged by remote: {"headerTableSize":4096,"enablePush":true,"initialWindowSize":65535,"maxFrameSize":16384,"maxConcurrentStreams":4294967295,"maxHeaderListSize":4294967295,"maxHeaderSize":4294967295,"enableConnectProtocol":false}
2022-10-07 09:43:33.907 CESTI1007 07:43:33.907648 1 dns.go:161] [dns] << recursed type=TXT name=my-service-name. rcode=NXDOMAIN answers=0 rtt=16.316482ms
2022-10-07 09:43:33.927 CESTI1007 07:43:33.926929 1 proxytransport.go:82] [proxy] end: POST url=https://my-service-name-qwsbralb2c-ez.a.run.app/MyService/SomeGrpcMethod took=21ms
2022-10-07 09:43:33.927 CEST2022/10/07 07:43:33 http: proxy error: x509: certificate signed by unknown authority
2022-10-07 09:43:33.927 CESTD 2022-10-07T07:43:33.927Z | call_stream | [1] Received server headers:
2022-10-07 09:43:33.928 CEST :status: 502
2022-10-07 09:43:33.928 CEST content-length: 0
2022-10-07 09:43:33.928 CEST date: Fri, 07 Oct 2022 07:43:33 GMT
2022-10-07 09:43:33.928 CESTD 2022-10-07T07:43:33.928Z | call_stream | [1] Received server trailers:
2022-10-07 09:43:33.928 CEST :status: 502
2022-10-07 09:43:33.928 CEST content-length: 0
2022-10-07 09:43:33.928 CEST date: Fri, 07 Oct 2022 07:43:33 GMT
2022-10-07 09:43:33.928 CESTD 2022-10-07T07:43:33.928Z | call_stream | [1] ended with status: code=14 details=""
2022-10-07 09:43:33.928 CESTD 2022-10-07T07:43:33.928Z | subchannel_refcount | (10) 127.0.0.1:80 callRefcount 1 -> 0
2022-10-07 09:43:33.928 CESTD 2022-10-07T07:43:33.928Z | call_stream | [1] close http2 stream with code 8
```

**Comment by bazaglia** [comment]
After some debugging, I concluded that everything looked fine on the gRPC client and the data sent to the proxy. I then installed the `ca-certificates` package to my Docker base image (Debian) and it fixed the problem! Curious that previously (when my Cloud Run services were public) I didn't need it even though the requests were also protected by HTTPS. Perhaps the certificates used by Google change when the Cloud Run service is private? Anyways, hope it helps someone with a similar issue.
